pkg/controller/statusmanager/pod_status.go

```diff
@@ -0,0 +1,278 @@
+package statusmanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	client "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// if a rollout has not made any progress by this time,
+	// mark ourselves as Degraded
+	ProgressTimeout = 10 * time.Minute
+
+	// lastSeenAnnotation - the annotation where we stash our state
+	lastSeenAnnotation = "network.operator.openshift.io/last-seen-state"
+)
+
+// podState is a snapshot of the last-seen-state and last-changed-times
+// for pod-creating entities
+type podState struct {
+	// "public" for marshalling to json, since we can't have complex keys
```

Comment:
It seems like it would be make more sense to just have a separate type for what gets json-marshaled since these fields are only used inside setLastPodState/getLastPodState?

Actually you don't even need a struct type for the internal type; just have the helper funcs take/return a pair of maps, and convert to/from a struct-of-arrays internally.

Reply:
yeah, good point.